nix/docker.nix

```diff
@@ -0,0 +1,141 @@
+#
+#   docker run -v /data -e NETWORK=mainnet inputoutput/cardano-explorer-api:<TAG>
+#   docker run -v /data -e NETWORK=mainnet inputoutput/cardano-tx-submit:<TAG>
+#
+# To launch with provided testnet configuration without persisting state run:
+#
+#   docker run -e NETWORK=testnet inputoutput/cardano-explorer-api:<TAG>
+#   docker run -e NETWORK=testnet inputoutput/cardano-tx-submit:<TAG>
+#
+# To launch with custom config, mount a dir containing config.json, genesis.json,
+# and topology,json into /config
+#
+#   docker run -v $PATH_TO/config:/config inputoutput/cardano-explorer-api:<TAG>
+#   docker run -v $PATH_TO/config:/config inputoutput/cardano-tx-submit:<TAG>
+#
+############################################################################
+
+{ iohkNix
+, commonLib
+, dockerTools
+
+# The main contents of the image.
+, cardano-explorer-api
+, cardano-tx-submit-webapi
+, scripts
+
+# Get the current commit
+, gitrev ? iohkNix.commitIdFromGitRepoOrZero ../.git
+
+# Other things to include in the image.
+, bashInteractive
+, cacert
+, coreutils
+, curl
+, glibcLocales
+, iana-etc
+, iproute
+, iputils
+, socat
+, utillinux
+, writeScriptBin
+, runtimeShell
+, lib
+
+, explorerApiRepoName ? "inputoutput/cardano-explorer-api"
+, txSubmitRepoName ? "inputoutput/cardano-tx-submit"
+}:
+
+let
+
+  # Layer of tools which aren't going to change much between versions.
+  baseImage = dockerTools.buildImage {
+    name = "base-env";
+    contents = [
+      bashInteractive   # Provide the BASH shell
+      cacert            # X.509 certificates of public CA's
+      coreutils         # Basic utilities expected in GNU OS's
+      curl              # CLI tool for transferring files via URLs
+      glibcLocales      # Locale information for the GNU C Library
+      iana-etc          # IANA protocol and port number assignments
+      iproute           # Utilities for controlling TCP/IP networking
+      iputils           # Useful utilities for Linux networking
+      socat             # Utility for bidirectional data transfer
+      utillinux         # System utilities for Linux
+    ];
+    # set up /tmp (override with TMPDIR variable)
+    extraCommands = ''
+      mkdir -m 0777 tmp
+    '';
+  };
+  apiDockerImage = let
+    entry-point = writeScriptBin "entry-point" ''
+      #!${runtimeShell}
+      #echo $NETWORK
+      if [[ -d /config ]]; then
+        echo PGPASSFILE = "/config/pgpass";
+        exec ${cardano-explorer-api}/bin/cardano-explorer-api
+      else
+        echo "Please place the pgpass file with credentials for postgres in /config"
+      fi
+    '';
+  in dockerTools.buildImage {
+    name = "${explorerApiRepoName}";
+    fromImage = baseImage;
+    tag = "${gitrev}";
+    created = "now";   # Set creation date to build time. Breaks reproducibility
+    contents = [ entry-point ];
+    config = {
+      EntryPoint = [ "${entry-point}/bin/entry-point" ];
+      ExposedPorts = {
+        "3001/tcp" = {};  # Cardano node p2p
```

Comment:
explorer-api doesnt use the cardano p2p protocol, but is exposing an http api on a port, dont remember which one

Reply:
8100